keps/sig-api-machinery/2211-strip-managed-fields-server-side/README.md

```diff
@@ -0,0 +1,426 @@
+required) or even code snippets. If there's any ambiguity about HOW your
+proposal will be implemented, this is the place to discuss them.
+-->
+
+### Server-side
+
+1. Add the three content types which represents the "unmanaged" format of objects to `k8s.io/apimachinery/pkg/runtime/types.go`:
+```go
+const (
+	ContentTypeJSON     string = "application/json"
+	ContentTypeYAML     string = "application/yaml"
+	ContentTypeProtobuf string = "application/vnd.kubernetes.protobuf"
+
+	ContentTypeUnmanagedJSON     string = "application/json+unmanaged"
+	ContentTypeUnmanagedYAML     string = "application/yaml+unmanaged"
+	ContentTypeUnmanagedProtobuf string = "application/vnd.kubernetes.protobuf+unmanaged"
+)
+```
+
+2. Add a new serializer to `k8s.io/apimachinery/pkg/runtime/serializer/` which wraps the existing serializers and
+strip the managedFields before encoding:
+```go
+type serializer struct {
+	inner      runtime.Serializer
+}
+
+func NewSerializer(s runtime.Serializer) runtime.Serializer {
+	if s == nil {
+		return nil
+	}
+	return &serializer{inner: s}
+}
+
+func (s *serializer) Encode(obj runtime.Object, w io.Writer) error {
+	if _, err := meta.Accessor(obj); err == nil {
+		obj = obj.DeepCopyObject()
+		a, _ := meta.Accessor(obj)
+		a.SetManagedFields(nil)
+	} else if meta.IsListType(obj) {
+		obj = obj.DeepCopyObject()
+		_ = meta.EachListItem(obj, func(item runtime.Object) error {
+			a, err := meta.Accessor(item)
+			if err != nil {
+				// ignore
+				return nil
+			}
+			a.SetManagedFields(nil)
+			return nil
+		})
+	}
+	return s.inner.Encode(obj, w)
+}
+
+func (s *serializer) Decode(data []byte, defaults *schema.GroupVersionKind, into runtime.Object) (runtime.Object, *schema.GroupVersionKind, error) {
+	return s.inner.Decode(data, defaults, into)
+}
+```
+
+3. Make `k8s.io/apimachinery/pkg/runtime/serializer.newSerializersForScheme` returns serializers for the new "unmanaged" content types.
+For instance:
+```go
+import (
+	...
+	"k8s.io/apimachinery/pkg/runtime/serializer/unmanaged"
+)
+
+func newSerializersForScheme(scheme *runtime.Scheme, mf json.MetaFactory, options CodecFactoryOptions) []serializerType {
+	...
+
+	yamlSerializer := json.NewSerializerWithOptions(
+		mf, scheme, scheme,
+		json.SerializerOptions{Yaml: true, Pretty: false, Strict: options.Strict},
+	)
+	yamlUnmanagedSerializer := unmanaged.NewSerializer(yamlSerializer)
+
+	serializers := []serializerType{
+		...
+		{
+			AcceptContentTypes: []string{runtime.ContentTypeYAML},
+			ContentType:        runtime.ContentTypeYAML,
+			FileExtensions:     []string{"yaml"},
+			EncodesAsText:      true,
+			Serializer:         yamlSerializer,
+		},
+		{
+			AcceptContentTypes: []string{runtime.ContentTypeUnmanagedYAML},
+			ContentType:        runtime.ContentTypeUnmanagedYAML,
+			FileExtensions:     []string{"yaml"},
+			EncodesAsText:      true,
+			Serializer:         yamlUnmanagedSerializer,
+		},
+	}
+}
+```
+
+### Client-side
```

Comment:
@julianvmodesto We'll try to get this KEP merged as provisional quickly, so that you can fill in that section if you're ok with it :-)

Reply:
The plan is to implement this using `Accept: application/yaml+unmanaged,application/yaml;q=0.9`, or simply: `Accept: application/yaml+unmanaged,application/yaml`

Reply:
what does unmanaged mean here ?

Reply:
> what does unmanaged mean here ?

Omit the managed fields.